src/formatter/formatter.vala

```diff
@@ -0,0 +1,278 @@
+using Gee;
+
+class Vls.Formatter : Object {
+    private Pair<Vala.SourceFile, Compilation> input;
+    private Lsp.FormattingOptions options;
+
+    public Formatter (Lsp.FormattingOptions options, Pair<Vala.SourceFile, Compilation> input) {
+        this.options = options;
+        this.input = input;
+    }
+
+    public string? format (out Lsp.TextEdit edit, out Jsonrpc.ClientError error) {
```

Comment:
I think you should have this function throw an error that you define. Then catch the error when this function is called.